src/kvstore/kvstore_dist.h

```diff
@@ -272,16 +282,35 @@ class KVStoreDist : public KVStoreLocal {
         // This shouldn't affect training of networks though because training involves
         // a sequence of push, pull, then push. This imposes ordering that the
         // second push happens after the first pull, and the pull happens after first push.
-        send_buf = merged;  // avoid memory copy
+        if (send_buf.is_none()) {
+          if (storage_type == kDefaultStorage) {
+            send_buf = NDArray(merged.shape(), pinned_ctx_, true, mshadow::DataType<DType>::kFlag);
+          } else {
+            send_buf = NDArray(storage_type, merged.shape(), pinned_ctx_,
+                               true, mshadow::DataType<DType>::kFlag);
+          }
+        }
+        if (merged.dtype() == mshadow::DataType<DType>::kFlag) {
+          send_buf = merged;
+        } else {
+          CopyFromTo(merged, &send_buf);
+        }
       } else {
         if (send_buf.is_none()) {
           if (storage_type == kDefaultStorage) {
-            send_buf = NDArray(merged.shape(), pinned_ctx_, true, merged.dtype());
+            send_buf = NDArray(merged.shape(), pinned_ctx_, true, mshadow::DataType<DType>::kFlag);
           } else {
-            send_buf = NDArray(storage_type, merged.shape(), pinned_ctx_, true, merged.dtype());
+            send_buf = NDArray(storage_type, merged.shape(), pinned_ctx_,
+                               true, mshadow::DataType<DType>::kFlag);
           }
         }
-        CopyFromTo(merged, &send_buf);
+        if (merged.dtype() == mshadow::DataType<DType>::kFlag) {
+          CopyFromTo(merged, &send_buf);
+        } else {
+          NDArray tmp = NDArray(merged.shape(), pinned_ctx_, true, merged.dtype());
```

Comment:
Here it looks like if the merged is on gpu, then it is copied to CPU. Why two copies? Both tmp and send_buf are on pinned_ctx, right? And if converting to fp16 is faster on GPU then why are we not doing that here?

Reply:
1. Copy two times, because it doesn't supports to copy from gpu with fp32 to cpu with fp16.  
2. Kvstore_dist just merge grads on all gpus in one machine.
Kvstore_dist_server merge the grads on all the machines.